// traceur/dist/commonjs/util-imports.js
"use strict";
var $__WebPageTranscoder_46_js__ = require("./WebPageTranscoder.js");
var $__HTMLImportTranscoder_46_js__ = require("./HTMLImportTranscoder.js");
Object.defineProperties(module.exports, {
  __esModule: {value: true},
  WebPageTranscoder: {
    enumerable: true,
    get: function() {
      return $__WebPageTranscoder_46_js__.WebPageTranscoder;
    }
  },
  HTMLImportTranscoder: {
    enumerable: true,
    get: function() {
      return $__HTMLImportTranscoder_46_js__.HTMLImportTranscoder;
    }
  },
  util: {
    enumerable: true,
    get: function() {
      return util;
    }
  }
});
require("./util/MutedErrorReporter.js");
var $__1 = require("./Options.js"),
    addOptions = $__1.addOptions,
    CommandOptions = $__1.CommandOptions,
    Options = $__1.Options;
var ErrorReporter = require("./util/ErrorReporter.js").ErrorReporter;
var CollectingErrorReporter = require("./util/CollectingErrorReporter.js").CollectingErrorReporter;
var util = {
  addOptions: addOptions,
  CommandOptions: CommandOptions,
  CollectingErrorReporter: CollectingErrorReporter,
  ErrorReporter: ErrorReporter,
  Options: Options
};
