// traceur/dist/commonjs/syntax/TokenType.js
"use strict";
Object.defineProperties(module.exports, {
  __esModule: {value: true},
  AMPERSAND: {
    enumerable: true,
    get: function() {
      return AMPERSAND;
    }
  },
  AMPERSAND_EQUAL: {
    enumerable: true,
    get: function() {
      return AMPERSAND_EQUAL;
    }
  },
  AND: {
    enumerable: true,
    get: function() {
      return AND;
    }
  },
  ARROW: {
    enumerable: true,
    get: function() {
      return ARROW;
    }
  },
  AT: {
    enumerable: true,
    get: function() {
      return AT;
    }
  },
  BACK_QUOTE: {
    enumerable: true,
    get: function() {
      return BACK_QUOTE;
    }
  },
  BANG: {
    enumerable: true,
    get: function() {
      return BANG;
    }
  },
  BAR: {
    enumerable: true,
    get: function() {
      return BAR;
    }
  },
  BAR_EQUAL: {
    enumerable: true,
    get: function() {
      return BAR_EQUAL;
    }
  },
  BREAK: {
    enumerable: true,
    get: function() {
      return BREAK;
    }
  },
  CARET: {
    enumerable: true,
    get: function() {
      return CARET;
    }
  },
  CARET_EQUAL: {
    enumerable: true,
    get: function() {
      return CARET_EQUAL;
    }
  },
  CASE: {
    enumerable: true,
    get: function() {
      return CASE;
    }
  },
  CATCH: {
    enumerable: true,
    get: function() {
      return CATCH;
    }
  },
  CLASS: {
    enumerable: true,
    get: function() {
      return CLASS;
    }
  },
  CLOSE_ANGLE: {
    enumerable: true,
    get: function() {
      return CLOSE_ANGLE;
    }
  },
  CLOSE_CURLY: {
    enumerable: true,
    get: function() {
      return CLOSE_CURLY;
    }
  },
  CLOSE_PAREN: {
    enumerable: true,
    get: function() {
      return CLOSE_PAREN;
    }
  },
  CLOSE_SQUARE: {
    enumerable: true,
    get: function() {
      return CLOSE_SQUARE;
    }
  },
  COLON: {
    enumerable: true,
    get: function() {
      return COLON;
    }
  },
  COMMA: {
    enumerable: true,
    get: function() {
      return COMMA;
    }
  },
  CONST: {
    enumerable: true,
    get: function() {
      return CONST;
    }
  },
  CONTINUE: {
    enumerable: true,
    get: function() {
      return CONTINUE;
    }
  },
  DEBUGGER: {
    enumerable: true,
    get: function() {
      return DEBUGGER;
    }
  },
  DEFAULT: {
    enumerable: true,
    get: function() {
      return DEFAULT;
    }
  },
  DELETE: {
    enumerable: true,
    get: function() {
      return DELETE;
    }
  },
  DO: {
    enumerable: true,
    get: function() {
      return DO;
    }
  },
  DOT_DOT_DOT: {
    enumerable: true,
    get: function() {
      return DOT_DOT_DOT;
    }
  },
  ELSE: {
    enumerable: true,
    get: function() {
      return ELSE;
    }
  },
  END_OF_FILE: {
    enumerable: true,
    get: function() {
      return END_OF_FILE;
    }
  },
  ENUM: {
    enumerable: true,
    get: function() {
      return ENUM;
    }
  },
  EQUAL: {
    enumerable: true,
    get: function() {
      return EQUAL;
    }
  },
  EQUAL_EQUAL: {
    enumerable: true,
    get: function() {
      return EQUAL_EQUAL;
    }
  },
  EQUAL_EQUAL_EQUAL: {
    enumerable: true,
    get: function() {
      return EQUAL_EQUAL_EQUAL;
    }
  },
  ERROR: {
    enumerable: true,
    get: function() {
      return ERROR;
    }
  },
  EXPORT: {
    enumerable: true,
    get: function() {
      return EXPORT;
    }
  },
  EXTENDS: {
    enumerable: true,
    get: function() {
      return EXTENDS;
    }
  },
  FALSE: {
    enumerable: true,
    get: function() {
      return FALSE;
    }
  },
  FINALLY: {
    enumerable: true,
    get: function() {
      return FINALLY;
    }
  },
  FOR: {
    enumerable: true,
    get: function() {
      return FOR;
    }
  },
  FUNCTION: {
    enumerable: true,
    get: function() {
      return FUNCTION;
    }
  },
  GREATER_EQUAL: {
    enumerable: true,
    get: function() {
      return GREATER_EQUAL;
    }
  },
  IDENTIFIER: {
    enumerable: true,
    get: function() {
      return IDENTIFIER;
    }
  },
  IF: {
    enumerable: true,
    get: function() {
      return IF;
    }
  },
  IMPLEMENTS: {
    enumerable: true,
    get: function() {
      return IMPLEMENTS;
    }
  },
  IMPORT: {
    enumerable: true,
    get: function() {
      return IMPORT;
    }
  },
  IN: {
    enumerable: true,
    get: function() {
      return IN;
    }
  },
  INSTANCEOF: {
    enumerable: true,
    get: function() {
      return INSTANCEOF;
    }
  },
  INTERFACE: {
    enumerable: true,
    get: function() {
      return INTERFACE;
    }
  },
  JSX_IDENTIFIER: {
    enumerable: true,
    get: function() {
      return JSX_IDENTIFIER;
    }
  },
  LEFT_SHIFT: {
    enumerable: true,
    get: function() {
      return LEFT_SHIFT;
    }
  },
  LEFT_SHIFT_EQUAL: {
    enumerable: true,
    get: function() {
      return LEFT_SHIFT_EQUAL;
    }
  },
  LESS_EQUAL: {
    enumerable: true,
    get: function() {
      return LESS_EQUAL;
    }
  },
  LET: {
    enumerable: true,
    get: function() {
      return LET;
    }
  },
  MINUS: {
    enumerable: true,
    get: function() {
      return MINUS;
    }
  },
  MINUS_EQUAL: {
    enumerable: true,
    get: function() {
      return MINUS_EQUAL;
    }
  },
  MINUS_MINUS: {
    enumerable: true,
    get: function() {
      return MINUS_MINUS;
    }
  },
  NEW: {
    enumerable: true,
    get: function() {
      return NEW;
    }
  },
  NO_SUBSTITUTION_TEMPLATE: {
    enumerable: true,
    get: function() {
      return NO_SUBSTITUTION_TEMPLATE;
    }
  },
  NOT_EQUAL: {
    enumerable: true,
    get: function() {
      return NOT_EQUAL;
    }
  },
  NOT_EQUAL_EQUAL: {
    enumerable: true,
    get: function() {
      return NOT_EQUAL_EQUAL;
    }
  },
  NULL: {
    enumerable: true,
    get: function() {
      return NULL;
    }
  },
  NUMBER: {
    enumerable: true,
    get: function() {
      return NUMBER;
    }
  },
  OPEN_ANGLE: {
    enumerable: true,
    get: function() {
      return OPEN_ANGLE;
    }
  },
  OPEN_CURLY: {
    enumerable: true,
    get: function() {
      return OPEN_CURLY;
    }
  },
  OPEN_PAREN: {
    enumerable: true,
    get: function() {
      return OPEN_PAREN;
    }
  },
  OPEN_SQUARE: {
    enumerable: true,
    get: function() {
      return OPEN_SQUARE;
    }
  },
  OR: {
    enumerable: true,
    get: function() {
      return OR;
    }
  },
  PACKAGE: {
    enumerable: true,
    get: function() {
      return PACKAGE;
    }
  },
  PERCENT: {
    enumerable: true,
    get: function() {
      return PERCENT;
    }
  },
  PERCENT_EQUAL: {
    enumerable: true,
    get: function() {
      return PERCENT_EQUAL;
    }
  },
  PERIOD: {
    enumerable: true,
    get: function() {
      return PERIOD;
    }
  },
  PLUS: {
    enumerable: true,
    get: function() {
      return PLUS;
    }
  },
  PLUS_EQUAL: {
    enumerable: true,
    get: function() {
      return PLUS_EQUAL;
    }
  },
  PLUS_PLUS: {
    enumerable: true,
    get: function() {
      return PLUS_PLUS;
    }
  },
  PRIVATE: {
    enumerable: true,
    get: function() {
      return PRIVATE;
    }
  },
  PROTECTED: {
    enumerable: true,
    get: function() {
      return PROTECTED;
    }
  },
  PUBLIC: {
    enumerable: true,
    get: function() {
      return PUBLIC;
    }
  },
  QUESTION: {
    enumerable: true,
    get: function() {
      return QUESTION;
    }
  },
  REGULAR_EXPRESSION: {
    enumerable: true,
    get: function() {
      return REGULAR_EXPRESSION;
    }
  },
  RETURN: {
    enumerable: true,
    get: function() {
      return RETURN;
    }
  },
  RIGHT_SHIFT: {
    enumerable: true,
    get: function() {
      return RIGHT_SHIFT;
    }
  },
  RIGHT_SHIFT_EQUAL: {
    enumerable: true,
    get: function() {
      return RIGHT_SHIFT_EQUAL;
    }
  },
  SEMI_COLON: {
    enumerable: true,
    get: function() {
      return SEMI_COLON;
    }
  },
  SLASH: {
    enumerable: true,
    get: function() {
      return SLASH;
    }
  },
  SLASH_EQUAL: {
    enumerable: true,
    get: function() {
      return SLASH_EQUAL;
    }
  },
  STAR: {
    enumerable: true,
    get: function() {
      return STAR;
    }
  },
  STAR_EQUAL: {
    enumerable: true,
    get: function() {
      return STAR_EQUAL;
    }
  },
  STAR_STAR: {
    enumerable: true,
    get: function() {
      return STAR_STAR;
    }
  },
  STAR_STAR_EQUAL: {
    enumerable: true,
    get: function() {
      return STAR_STAR_EQUAL;
    }
  },
  STATIC: {
    enumerable: true,
    get: function() {
      return STATIC;
    }
  },
  STRING: {
    enumerable: true,
    get: function() {
      return STRING;
    }
  },
  SUPER: {
    enumerable: true,
    get: function() {
      return SUPER;
    }
  },
  SWITCH: {
    enumerable: true,
    get: function() {
      return SWITCH;
    }
  },
  TEMPLATE_HEAD: {
    enumerable: true,
    get: function() {
      return TEMPLATE_HEAD;
    }
  },
  TEMPLATE_MIDDLE: {
    enumerable: true,
    get: function() {
      return TEMPLATE_MIDDLE;
    }
  },
  TEMPLATE_TAIL: {
    enumerable: true,
    get: function() {
      return TEMPLATE_TAIL;
    }
  },
  THIS: {
    enumerable: true,
    get: function() {
      return THIS;
    }
  },
  THROW: {
    enumerable: true,
    get: function() {
      return THROW;
    }
  },
  TILDE: {
    enumerable: true,
    get: function() {
      return TILDE;
    }
  },
  TRUE: {
    enumerable: true,
    get: function() {
      return TRUE;
    }
  },
  TRY: {
    enumerable: true,
    get: function() {
      return TRY;
    }
  },
  TYPEOF: {
    enumerable: true,
    get: function() {
      return TYPEOF;
    }
  },
  UNSIGNED_RIGHT_SHIFT: {
    enumerable: true,
    get: function() {
      return UNSIGNED_RIGHT_SHIFT;
    }
  },
  UNSIGNED_RIGHT_SHIFT_EQUAL: {
    enumerable: true,
    get: function() {
      return UNSIGNED_RIGHT_SHIFT_EQUAL;
    }
  },
  VAR: {
    enumerable: true,
    get: function() {
      return VAR;
    }
  },
  VOID: {
    enumerable: true,
    get: function() {
      return VOID;
    }
  },
  WHILE: {
    enumerable: true,
    get: function() {
      return WHILE;
    }
  },
  WITH: {
    enumerable: true,
    get: function() {
      return WITH;
    }
  },
  YIELD: {
    enumerable: true,
    get: function() {
      return YIELD;
    }
  }
});
var AMPERSAND = '&';
var AMPERSAND_EQUAL = '&=';
var AND = '&&';
var ARROW = '=>';
var AT = '@';
var BACK_QUOTE = '`';
var BANG = '!';
var BAR = '|';
var BAR_EQUAL = '|=';
var BREAK = 'break';
var CARET = '^';
var CARET_EQUAL = '^=';
var CASE = 'case';
var CATCH = 'catch';
var CLASS = 'class';
var CLOSE_ANGLE = '>';
var CLOSE_CURLY = '}';
var CLOSE_PAREN = ')';
var CLOSE_SQUARE = ']';
var COLON = ':';
var COMMA = ',';
var CONST = 'const';
var CONTINUE = 'continue';
var DEBUGGER = 'debugger';
var DEFAULT = 'default';
var DELETE = 'delete';
var DO = 'do';
var DOT_DOT_DOT = '...';
var ELSE = 'else';
var END_OF_FILE = 'End of File';
var ENUM = 'enum';
var EQUAL = '=';
var EQUAL_EQUAL = '==';
var EQUAL_EQUAL_EQUAL = '===';
var ERROR = 'error';
var EXPORT = 'export';
var EXTENDS = 'extends';
var FALSE = 'false';
var FINALLY = 'finally';
var FOR = 'for';
var FUNCTION = 'function';
var GREATER_EQUAL = '>=';
var IDENTIFIER = 'identifier';
var IF = 'if';
var IMPLEMENTS = 'implements';
var IMPORT = 'import';
var IN = 'in';
var INSTANCEOF = 'instanceof';
var INTERFACE = 'interface';
var JSX_IDENTIFIER = 'jsx identifier';
var LEFT_SHIFT = '<<';
var LEFT_SHIFT_EQUAL = '<<=';
var LESS_EQUAL = '<=';
var LET = 'let';
var MINUS = '-';
var MINUS_EQUAL = '-=';
var MINUS_MINUS = '--';
var NEW = 'new';
var NO_SUBSTITUTION_TEMPLATE = 'no substitution template';
var NOT_EQUAL = '!=';
var NOT_EQUAL_EQUAL = '!==';
var NULL = 'null';
var NUMBER = 'number literal';
var OPEN_ANGLE = '<';
var OPEN_CURLY = '{';
var OPEN_PAREN = '(';
var OPEN_SQUARE = '[';
var OR = '||';
var PACKAGE = 'package';
var PERCENT = '%';
var PERCENT_EQUAL = '%=';
var PERIOD = '.';
var PLUS = '+';
var PLUS_EQUAL = '+=';
var PLUS_PLUS = '++';
var PRIVATE = 'private';
var PROTECTED = 'protected';
var PUBLIC = 'public';
var QUESTION = '?';
var REGULAR_EXPRESSION = 'regular expression literal';
var RETURN = 'return';
var RIGHT_SHIFT = '>>';
var RIGHT_SHIFT_EQUAL = '>>=';
var SEMI_COLON = ';';
var SLASH = '/';
var SLASH_EQUAL = '/=';
var STAR = '*';
var STAR_EQUAL = '*=';
var STAR_STAR = '**';
var STAR_STAR_EQUAL = '**=';
var STATIC = 'static';
var STRING = 'string literal';
var SUPER = 'super';
var SWITCH = 'switch';
var TEMPLATE_HEAD = 'template head';
var TEMPLATE_MIDDLE = 'template middle';
var TEMPLATE_TAIL = 'template tail';
var THIS = 'this';
var THROW = 'throw';
var TILDE = '~';
var TRUE = 'true';
var TRY = 'try';
var TYPEOF = 'typeof';
var UNSIGNED_RIGHT_SHIFT = '>>>';
var UNSIGNED_RIGHT_SHIFT_EQUAL = '>>>=';
var VAR = 'var';
var VOID = 'void';
var WHILE = 'while';
var WITH = 'with';
var YIELD = 'yield';
